app/src/main/java/com/sdp/movemeet/DistanceCalculator.java

```diff
@@ -1,68 +1,130 @@
+                activities.add(pair.getKey());
+            }
+
             return activities;
         } else {
             return null;
         }
     }
 
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
     public ArrayList<Activity> getTopActivities(Integer n) {
         if (sorted) {
-            ArrayList<Activity> sublist = new ArrayList<Activity> (activities.subList(0, n));
-            return sublist;
+            Stream<Activity> topActivitiesStream = activityDistanceMap.subList(0, n).stream().map(Pair::getKey);
+            return topActivitiesStream.collect(Collectors.toCollection(ArrayList::new));
+
         } else {
             return null;
         }
     }
 
-    /*
     public ArrayList<Activity> getActivitiesInRadius(Double radius) {
-        return null;
-    }
-    */
+        if (sorted) {
+            ArrayList<Activity> topActivities = new ArrayList<Activity>();
 
-    @RequiresApi(api = Build.VERSION_CODES.N)
-    public void sortActivities() {
-        if (activities.isEmpty()) { // == null
-            sorted = false;
+            int i = 0;
+            //System.out.println(activityDistanceMap.get(i).getValue());
+            while (activityDistanceMap.get(i).getValue() <= radius) {
+                topActivities.add(activityDistanceMap.get(i).getKey());
+                i++;
+            }
+            return topActivities;
         } else {
-            activities.sort((Activity a1, Activity a2) ->
-            calculateDistance(userLatitude, userLongitude, a1.getLatitude(), a1.getLongitude())
-            .compareTo(
-            calculateDistance(userLatitude, userLongitude, a2.getLatitude(), a2.getLongitude())));
-            sorted = true;
+            return null;
         }
     }
 
+
+    public void calculateDistances() {
+        for (Pair pair : activityDistanceMap) {
+            pair.setValue(calculateDistance(
+                    userLatitude, userLongitude,
+                    pair.getKey().getLatitude(), pair.getKey().getLongitude()
+            ));
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    public void sort() {
+        activityDistanceMap.sort(new Comparator<Pair>() {
+            @Override
+            public int compare(Pair o1, Pair o2) {
+                return o1.getValue().compareTo(o2.getValue());
+            }
+        });
+
+        sorted = true;
+    }
+
     public boolean isSorted() {
         return sorted;
     }
 
     public Double calculateDistance(Double lat1, Double lng1, Double lat2, Double lng2) {
-        return 2 * EARTH_RADIUS * Math.asin(Math.sqrt(Math.pow(Math.sin((lat2 - lat1)/2),2) + Math.cos(lat1) * Math.cos(lat2) * Math.pow(Math.sin((lng2 - lng1)/2),2)));
+        // Double p = 0.017453292519943295;
+        Double p = Math.PI / 180;
+
+        //Double a =  Math.pow(Math.sin((lat2 - lat1)/2),2) + Math.cos(lat1) * Math.cos(lat2) * Math.pow(Math.sin((lng2 - lng1)/2),2);
+        Double a = 0.5 - Math.cos((lat2 - lat1) * p)/2 +
```

Comment:
What is the purpose of the two commented lines?

Reply:
The first constant was just a hardcoded PI/180.
I decided to replace it with the explicit expression rather than a "magic number" so that it makes more sense.

The second comment was for the previously used, slow and unoptimized formula. It is now useless.

Thanks for the question, this is essentially dead code and I shall remove these comments in my next commit.